File: app/src/main/java/com/annapol04/munchkin/engine/Card.java
package com.annapol04.munchkin.engine;

import android.support.annotation.DrawableRes;
import android.support.annotation.StringRes;

import java.util.ArrayList;
import java.util.List;

public class Card {
    private static final List<Card> lookup = new ArrayList<>();
    private static int idSource = 0;

    private final int id;
    private final @StringRes int name;
    private final @StringRes int description;
    private final @DrawableRes int imageResourceID;

    public Card(@StringRes int name, @StringRes int description, @DrawableRes int imageResourceID) {
        this.id = idSource++;
        this.name = name;
        this.description = description;
        this.imageResourceID = imageResourceID;

        lookup.add(this);
    }

    public int getId() {
        return id;
    }

    public int getName() {
        return this.name;
    }

    public int getImageResourceID() {
        return imageResourceID;
    }

    public static Card fromId(int id) {
        if (id >= lookup.size())
            throw new IllegalArgumentException("Invalid event card id " + id);

        return lookup.get(id);
    }
}

File: app/src/main/java/com/annapol04/munchkin/engine/Card.kt
package com.annapol04.munchkin.engine

import android.support.annotation.DrawableRes
import android.support.annotation.StringRes

import java.util.ArrayList

open class Card(@param:StringRes @field:StringRes
                val name: Int,
                @param:StringRes @field:StringRes
                private val description: Int,
                @param:DrawableRes @field:DrawableRes
                val imageResourceID: Int) {

    val id: Int

    init {
        this.id = idSource++

        lookup.add(this)
    }

    companion object {
        private val lookup = ArrayList<Card>()
        private var idSource = 0

        fun fromId(id: Int): Card {
            if (id >= lookup.size)
                throw IllegalArgumentException("Invalid event card id $id")

            return lookup[id]
        }
    }

}
